Reset password frontend

PatricKangasmaki opened the pull request:
Some notes:
- Email view is currently accessible in /resetPasswordEmail/:token
- Email view still has things marked as to do since it will most likely need changes when actually implementing it to the email
- All of the pages are using old footers but I figured that creating new footers and applying it to all the other pages as well (email verification etc.) is not in the scope of this PR

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,152 @@
+      if (disableTime > 0) setDisableTime((previous) => previous - 1);
+    }, 1000);
+
+    setTimeout(() => setResendDisabled(false), defaultDisableTime * 1000);
+    setTimeout(() => {
+      clearInterval(countdownInterval);
```
```suggestion
    setTimeout(() => {
      setResendDisabled(false);
      clearInterval(countdownInterval);
```

Couldn't these be in the same timeout?

The timeout should also be cancelled when the component is unmounted.

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,152 @@
+  const [resendDisabled, setResendDisabled] = useState(false);
+  const [disableTime, setDisableTime] = useState(defaultDisableTime);
+
+  const sendEmailLink = () => {
+    api.sendPasswordResetLink(email);
+  };
```
The possible errors should be checked. For example unknown email causes 404, but the page shows as if all went fine.

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,152 @@
+    disableSending();
+  };
+
+  const sendLinkView = () => {
+    return (
```
```suggestion
  const sendLinkView = () => (
```

This is somewhat a personal preference, but I find the single expression form nicer as it reduces the indentation

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,152 @@
+              </button>
+            </div>
+            <div>
+              {!resendDisabled && (
+                <>
+                  <Trans i18nKey="Didn’t receive the email" />{' '}
+                  <button
+                    type="button"
+                    className={classes(css.linkButton, css.green)}
+                    onClick={handleResend}
+                  >
+                    <Trans i18nKey="Resend link" />
+                  </button>
+                </>
+              )}
+              {resendDisabled && (
```
`resendDisabled ? a : b` would be clearer

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,152 @@
+
+  return (
+    <>
+      {view === 'Send link' && sendLinkView()}
+      {view === 'Link sent' && linkSentView()}
```
```suggestion
      {linkSent ? linkSentView() : sendLinkView()}
```

The `view` would make sense as a boolean (or typed enum e.g. `useState<'a' | 'b' | 'c'>('a')` if there is more than two options)

juhoeerola commented on frontend/src/pages/ResetPasswordPage.tsx:
```diff
@@ -0,0 +1,160 @@
+
+  return (
+    <>
+      {view === 'Reset password' && resetPasswordView()}
+      {view === 'Done' && doneView()}
```
```suggestion
      {done ? doneView() : resetPasswordView()}
```

juhoeerola commented on frontend/src/pages/ResetPasswordPage.tsx:
```diff
@@ -0,0 +1,160 @@
+    });
+  };
+
+  const resetPasswordView = () => {
+    return (
+      <>
+        <div className="formDiv">
```
```suggestion
  const resetPasswordView = () => (
      <div className="formDiv">
```

At least the unnecessary fragment should be removed

juhoeerola commented on frontend/src/pages/ResetPasswordEmail.tsx:
```diff
@@ -0,0 +1,73 @@
+
+const classes = classNames.bind(css);
+
+export const ResetPasswordEmail = () => {
```
I don't think we need the "show email in browser" option. Or do we?

Anyway, I would rather have this as a separate pr.

juhoeerola reviewed (changes requested):
The flow feels real nice 👌 
One thing I noticed though, is that clicking "send link" -> "change address" -> "send link" -> ... allows skipping the cooldown.


I think the email page should be in separate pr to keep this one a bit more manageable and focused.

PatricKangasmaki commented:
Great points 👍 Added all your suggestions

> One thing I noticed though, is that clicking "send link" -> "change address" -> "send link" -> ... allows skipping the cooldown.

Added separate cooldown for the send link button on the first page. Made it 15 sec so it prevents spamming but shouldn't affect real cases where you accidentally typed wrong address too much.

> I think the email page should be in separate pr to keep this one a bit more manageable and focused.  

Removed the email page from this pr

> The timeout should also be cancelled when the component is unmounted.  

Missed this part in the previous fixes, it's under progress now.

IidaKainu commented on frontend/src/pages/ResetPasswordPage.tsx:
```diff
@@ -0,0 +1,153 @@
+
+    if (!token) return;
+
+    dispatch(userActions.resetPassword({ token, password })).then((res) => {
+      if (userActions.resetPassword.rejected.match(res)) {
+        if (res.payload?.response?.status === 403)
+          setErrorMessage(t('Token expired'));
+        else if (res.payload?.response?.status === 404)
+          setErrorMessage(t('Token not found'));
+        else if (res.payload) {
+          setErrorMessage(t('Error message', { error: res.payload.message }));
+        }
+      } else {
+        setDone(true);
+      }
+    });
```
The nesting here could be reduced (generally makes code easier to understand). Inverting an "if" statement is a great way achieve it. [Example](https://www.refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html)
```suggestion
    const res = await dispatch(userActions.resetPassword({ token, password }));
    if (!userActions.resetPassword.rejected.match(res)) {
      setDone(true);
      return;
    }
    if (res.payload?.response?.status === 403)
      setErrorMessage(t('Token expired'));
    else if (res.payload?.response?.status === 404)
      setErrorMessage(t('Token not found'));
    else if (res.payload)
      setErrorMessage(t('Error message', { error: res.payload.message }));
```

PatricKangasmaki replied:
Good point, your version looks a lot cleaner when there's no 3 nested if's 👍 Added the suggestion

IidaKainu reviewed (changes requested):
I have to agree the user flow is great 🔥

IidaKainu reviewed (approved):
Approved on my behalf 👍

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,251 @@
+      return;
+    }
+
+    setResendDisableTime(defaultDisableTime / 2);
+    createDisableTimer(
```
```suggestion
    createDisableTimer(
```

`createDisableTime()` could set the resend disable time, since the `time` parameter is just that

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,251 @@
+
+  const [linkSent, setLinkSent] = useState(false);
+  const [email, setEmail] = useState('');
+  const [sendDisabled, setSendDisabled] = useState(false);
+  const [sendDisableTime, setSendDisableTime] = useState(
+    defaultDisableTime / 2,
+  );
+  const [resendDisabled, setResendDisabled] = useState(false);
+  const [resendDisableTime, setResendDisableTime] = useState(
+    defaultDisableTime,
+  );
```
I think the two flags & times should use just one and same flag & time. Just setting the time to different initial value from different places.

This would simplify the code and keep the timer consistent when clicking to "change address" after resending.

PatricKangasmaki replied:
This continues from the above explanation. The sendDisableTime can be going 15->14->13->12 etc. while the resendDisableTime is going from 30->29->... so they needed their own times in this particular implementation (at least I think they are needed).

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,251 @@
+
+export const ForgotPasswordPage = () => {
+  const { t } = useTranslation();
+  const defaultDisableTime = 30; // Seconds
```
```suggestion
  const defaultDisableSeconds = 30;
```

The comment could be moved to the name, then it would also be visible in each use

juhoeerola reviewed (changes requested):
Mostly about the timeout handling, and couple other minor suggestions.

For the other parts looks good to me.

PatricKangasmaki commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,251 @@
+  const [errorMessage, setErrorMessage] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+
+  let timeoutArray = useMemo(() => {
+    const array: NodeJS.Timeout[] = [];
+    return array;
+  }, []);
+
+  let intervalArray = useMemo(() => {
+    const array: NodeJS.Timeout[] = [];
+    return array;
+  }, []);
+
+  // Cleanup timers after unmounting
+  useEffect(
+    () => () => {
+      timeoutArray.forEach((timeout) => {
+        clearTimeout(timeout);
+      });
+      intervalArray.forEach((interval) => {
+        clearInterval(interval);
+      });
+    },
+    [timeoutArray, intervalArray],
+  );
```
Do you mean generally or in this specific scenario? My reasoning for the two different timers here is that I made them to serve different purposes.

The first timeout is for setting timer to the "Send link" -button on the sendLinkView. It starts when user requests a link on that first view. The reason for this timeout is that you can't spam emails by sending link -> change address -> send link etc. If user is using this page correctly, they should never even know about this timeout, that's why it's 15 seconds instead of the 30 which is on the second view as 30 seconds could have effect on normal use as well eg. accidentally entering wrong email and realizing it in 10 seconds and then having to wait another 20. This is just a small protection against malicious use.

The second timeout is for the "Resend link" -button on the linkSentView. It starts when users requests a new link (after the first email didn't supposedly arrive). The reason for this being different timer than the sendLink one is that it didn't look as good and smooth (to me) if the "Didn’t receive the email? Resend link" was already disabled when entering the page the first time.  

And that's why both of these timers could be active at the same time since they get set from different actions and I had to clean them both in the cleanup

PatricKangasmaki replied:
Maybe I'm thinking about these timers way too complicatedly and the pages could just share the timers 😬

juhoeerola replied:
> the pages could just share the timers

This is what I was thinking. It would also keep the time consistent when the user navigates to change email during the 30 s timer.

> This is just a small protection against malicious use.

Any enforcement of the send rate must be done on the server side anyway, but it's good that the frontend indicates that the user should not spam the button if they didn't receive the email in a heartbeat.

PatricKangasmaki commented:
Added all the changes 👍 There's now just one timer which controls both send and resend buttons

juhoeerola commented on frontend/src/pages/ForgotPasswordPage.tsx:
```diff
@@ -0,0 +1,212 @@
+      return;
+    }
+
+    createDisableTimer();
+    setErrorMessage('');
+    setIsLoading(false);
+    setLinkSent(true);
+  };
+
+  const handleResend = () => {
+    sendEmailLink();
+    createDisableTimer();
```
```suggestion
    createDisableTimer(defaultDisableSeconds / 2);
    setErrorMessage('');
    setIsLoading(false);
    setLinkSent(true);
  };

  const handleResend = () => {
    sendEmailLink();
    createDisableTimer(defaultDisableSeconds);
```

The time could still be a parameter, so typo in email isn't quite as long wait

juhoeerola reviewed (approved):
Looks good 👍 

The time could be parametrised, but either way is fine, your call

PatricKangasmaki commented:
> The time could be parametrised, but either way is fine, your call

Added this 👍
